wire-ios-sync-engine/Source/SessionManager/UpdateEventMigratorDAO.swift

```diff
@@ -0,0 +1,231 @@
+import WireAPI
+
+// sourcery: AutoMockable
+protocol UpdateEventMigratorDAOProtocol {
+
+    func existsLegacyEvent() async throws -> Bool
+    func indexOfLastEventEnvelope() async throws -> Int64
+    func nextBatchOfLegacyEvents(privateKeys: EARPrivateKeys?) async -> [ZMUpdateEvent]?
+    func insertEventEnvelope(_ eventEnvelope: UpdateEventEnvelope, index: Int64) async throws
+    func deleteNextBatchOfLegacyEvents() async
+    func save() async throws
+
+}
+
+struct UpdateEventMigratorDAO: UpdateEventMigratorDAOProtocol {
+
+    private let context: NSManagedObjectContext
+    private let encoder = JSONEncoder()
+
+    init(context: NSManagedObjectContext) {
+        self.context = context
+    }
+
+    func existsLegacyEvent() async throws -> Bool {
+        try await context.perform {
+            let request = StoredUpdateEvent.fetchRequest()
+            return try context.count(for: request) > 0
+        }
+    }
+
+    func indexOfLastEventEnvelope() async throws -> Int64 {
+        try await context.perform {
+            let request = StoredUpdateEventEnvelope.lastObjectFetchRequest
+            let lastEnvelope = try context.fetch(request).first
+            return lastEnvelope?.sortIndex ?? 0
+        }
+    }
+
+    func nextBatchOfLegacyEvents(privateKeys: EARPrivateKeys?) async -> [ZMUpdateEvent]? {
+        await context.perform {
+            let legacyEvents = StoredUpdateEvent.nextEvents(
+                context,
+                batchSize: 500,
+                callEventsOnly: false
+            )
+            .compactMap {
+                switch StoredUpdateEvent.extractUpdateEvent(
+                    from: $0,
+                    privateKeys: privateKeys
+                ) {
+                case let .success(legacyEvent):
+                    return legacyEvent
+
+                case .failure:
+                    return nil
+                }
+            }
+
+            guard !legacyEvents.isEmpty else {
+                return nil
+            }
+
+            return legacyEvents
+        }
+    }
+
+    func insertEventEnvelope(
+        _ eventEnvelope: UpdateEventEnvelope,
+        index: Int64
+    ) async throws {
+        try await context.perform {
+            _ = StoredUpdateEventEnvelope.insertNewObject(
+                data: try encoder.encode(eventEnvelope),
+                sortIndex: index,
+                in: context
+            )
+        }
+    }
+
+    func deleteNextBatchOfLegacyEvents() async {
+        await context.perform {
+            StoredUpdateEvent.nextEvents(
+                context,
+                batchSize: 500,
+                callEventsOnly: false
+            ).forEach {
+                context.delete($0)
+            }
+        }
+    }
+
+    func save() async throws {
+        try await context.perform {
+            try context.save()
+        }
+    }
+
+}
+
+@available(iOS 17, *)
+actor UpdateEventMigratorDAO2: UpdateEventMigratorDAOProtocol {
```

Comment:
I'm wondering if I should leave this here or remove it and use the other implementation instead.